subprojects/test-runner/client/src/main/kotlin/com/avito/runner/scheduler/runner/scheduler/TestSchedulerFactoryImpl.kt

```diff
@@ -1,47 +1,229 @@
 
-    override fun create(): TestScheduler {
-        val testSuiteProvider: TestSuiteProvider = createTestSuiteProvider()
+    override fun create(params: RunnerInputParams): TestScheduler {
+
+        val timeProvider: TimeProvider = DefaultTimeProvider()
+
+        val statsdSender = createStatsdSender(params)
+
+        val httpClientProvider = HttpClientProvider(
+            statsDSender = statsdSender,
+            timeProvider = timeProvider,
+            loggerFactory = params.loggerFactory
+        )
+
+        val reportFactory: ReportFactory = createReportFactory(
+            params = params,
+            timeProvider = timeProvider,
+            httpClientProvider = httpClientProvider
+        )
+
+        val report = reportFactory.createReport()
+
+        val filterFactory = createFilterFactory(
+            params = params,
+            reportFactory = reportFactory
+        )
+
+        val testSuiteProvider: TestSuiteProvider = createTestSuiteProvider(
+            params = params,
+            report = report,
+            filterFactory = filterFactory
+        )
+
+        val metricsConfig: RunnerMetricsConfig = createRunnerMetricsConfig(params)
+
+        val finalizerFactory: FinalizerFactory = createFinalizerFactory(
+            params = params,
+            metricsConfig = metricsConfig,
+            reportFactory = reportFactory,
+            timeProvider = timeProvider
+        )
+
+        val androidDebugBridgeProvider = createAndroidDebugBridgeProvider(params.loggerFactory)
+
+        val emulatorsLogsReporterProvider = createEmulatorsLogsReporterProvider(params)
+
+        val testRunnerFactoryProvider = createTestRunnerFactoryProvider(
+            params = params,
+            timeProvider = timeProvider,
+            httpClientProvider = httpClientProvider,
+            report = report,
+            metricsConfig = metricsConfig,
+            androidDebugBridgeProvider = androidDebugBridgeProvider,
+            emulatorsLogsReporterProvider = emulatorsLogsReporterProvider
+        )
 
         return TestSchedulerImpl(
             params = params,
             report = report,
             testSuiteProvider = testSuiteProvider,
-            testSuiteLoader = testSuiteLoader,
+            testSuiteLoader = createTestSuiteLoader(),
             filterInfoWriter = FilterInfoWriter.Impl(
                 outputDir = params.outputDir,
             ),
             loggerFactory = params.loggerFactory,
             testRunnerFactory = testRunnerFactoryProvider.provide(),
+            finalizer = finalizerFactory.create(),
         )
     }
 
-    private fun createTestSuiteProvider(): TestSuiteProvider = TestSuiteProvider.Impl(
-        report = report,
-        targets = params.instrumentationConfiguration.targets,
-        reportSkippedTests = params.instrumentationConfiguration.reportSkippedTests,
-        filterFactory = FilterFactory.create(
+    private fun createFinalizerFactory(
```

Comment:
`TestShedulerFactory` must not know how to create a Finalizer
Make FinalizerFactory as an input parameter

Reply:
Finalizer is now hidden in TestScheduler implementation, so i can only extract it like this:

```
TestSchedulerFactoryImpl(finalizerFactory (params) -> Filnalizer) {

   override fun create(params: RunnerInputParams): TestScheduler {
        val finalizerFactory = finalizerFactory.invoke(params)
   }
}
```

which is odd